### ISSUE TITLE
Firestore and Auth cause crash after change in version

### ISSUE THREAD

**Issue by Jacobsjj2** [firebase/quickstart-unity#905]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2019.4.16f1
  * Firebase Unity SDK version: 7.0.1
  * Source you installed the SDK: .unitypackage
  * Problematic Firebase Component: Auth, Firestore
  * Other Firebase Components in use: Analytics
  * Additional SDKs you are using: N/A
  * Platform you are using the Unity editor on: Linux
  * Platform you are targeting: Android
  * Scripting Runtime: Mono

### [REQUIRED] Please describe the issue here:

I downgraded from SDK 7.0.0 to 6.16.1 due to a bug where unicode characters weren't showing up. My downgrade process was simply deleting the firebase plugins using the dependency manager and then importing the .unitypackage files again and resolving Android dependencies. After running the app again whenever I get to a screen that accesses auth or firestore functions the Unity editor crashes with no error messages being printed to the console. I surmised it might be an issue with the version so to make sure I switched back to 7.0.0 and finally upgraded to 7.0.1 but with the same results.

`Fallback handler could not load library /home/jj/Programs/Unity/2019.4.16f1/Editor/Data/Mono/lib/libAssets/Firebase/Plugins/x86_64/FirebaseCppApp-7_0_1.so`

Full log available [here](https://gist.github.com/Jacobsjj2/cd7fd711e8086b86add330b7261cb065).

In the Unity logs however there seems to be a file being called that doesn't exist. This is similar to issues #590, #628 and #371 but unfortunately I don't seem to have the same problems and most were supposedly fixed with upgrading.

#### Steps to reproduce:

I suspect this is either a result of the change in version or maybe a separate error occurring after some time using the plugins, but on my machine using the below attached script on any simple gameObject and adding random buttons and an input field to the serializable assets will cause a crash on my end.

#### Relevant Code:

```
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;
using Firebase.Firestore;

public class SettingsMenu : MonoBehaviour
{
    [SerializeField] private InputField nameTextField;
    [SerializeField] private Button loginButton;
    [SerializeField] private Button logoutButton;
    Firebase.Auth.FirebaseAuth auth;
    Firebase.Firestore.ListenerRegistration userTableListener;
    private string savedName;

    void Awake()
    {
        auth = Firebase.Auth.FirebaseAuth.DefaultInstance;
        auth.StateChanged += ListenForAuthChange;
    }

    void ListenForSettingsChange()
    {
        string userId = getUserId();
        if (userTableListener != null)
        {
            userTableListener.Stop();
        }
        if (userId == null)
        {
            return;
        }

        FirebaseFirestore db = FirebaseFirestore.DefaultInstance;
        DocumentReference userRef = db.Collection("users").Document(userId);
        userTableListener = userRef.Listen(snapshot =>
        {
            Dictionary<string, object> userSettings = snapshot.ToDictionary();
            if (!userSettings.ContainsKey("userName"))
            {
                return;
            }
            object userNameField = userSettings["userName"];

            if (userNameField == null)
            {
                return;
            }
            else
            {
                nameTextField.text = userNameField.ToString();
                savedName = userNameField.ToString();
            }
        });
    }

    private void ListenForAuthChange(object sender, System.EventArgs eventArgs)
    {
        if (auth.CurrentUser == null)
        {
            logoutButton.gameObject.SetActive(false);
            loginButton.gameObject.SetActive(true);
            nameTextField.interactable = false;
        }
        else
        {
            loginButton.gameObject.SetActive(false);
            logoutButton.gameObject.SetActive(true);
            nameTextField.interactable = true;
        }
        ListenForSettingsChange();
    }

    private string getUserId()
    {
        Firebase.Auth.FirebaseUser user = auth.CurrentUser;
        if (user == null)
        {
            return null;
        }
        else
        {
            return user.UserId;
        }
    }

    public void Logout()
    {
         auth.SignOut();
    }
}

```

**Comment by vimanyu** [comment]
Hi @Jacobsjj2,
Can you please try Unity plugins 7.0.2 and check if you are seeing this there too? They have a fix for the unicode firestore issue.

**Comment by Jacobsjj2** [comment]
> Hi @Jacobsjj2,
> Can you please try Unity plugins 7.0.2 and check if you are seeing this there too? They have a fix for the unicode firestore issue.

Hi @vimanyu, thank you for letting me know about the new version! Unfortunately this still crashes my app. But I've been able to track down the crash-causing code to only the snapshot part:

```
FirebaseFirestore db = FirebaseFirestore.DefaultInstance;
DocumentReference userRef = db.Collection("users").Document(userId);
userTableListener = userRef.Listen(snapshot =>
{
        Dictionary<string, object> userSettings = snapshot.ToDictionary();
        if (!userSettings.ContainsKey("userName"))
        {
            return;
        }
        object userNameField = userSettings["userName"];

        if (userNameField == null)
        {
            return;
        }
        else
        {
            nameTextField.text = userNameField.ToString();
            savedName = userNameField.ToString();
        }
});
```

It seems to execute fine but crashes the moment a snapshot comes through. I've added logs to the start of the method to try to see if the contents of the callback are causing the problem but no logs appear so I know the callback content is never executed. I also replaced `userId` with a random String to make sure it's not a null String error but all with the same crash unfortunately.

**Comment by Jacobsjj2** [comment]
The following commands also cause a crash so I believe this is specifically related to Firestore:

```
db.TerminateAsync();
db.ClearPersistenceAsync();
```

**Comment by Jacobsjj2** [comment]
I was able to solve my problem just now when I referred back to the log and noticed this line:

```
/home/jj/Programs/Unity/2019.4.16f1/Editor/Data/Resources/PackageManager/Server/node_modules/winston/lib/winston/logger.js:307
      throw ex;
```
This confused me initially because it occurs after the log indicates it's launching the bug reporter. I had recently installed the firebase cli tools elsewhere and in the process installed a newer version of node.js. After searching around I found [this post](https://stackoverflow.com/questions/51435990/what-is-this-error-in-winston) which suggested running 

```
npm install npm -g
```
This actually downgraded npm (?) but it did fix my issue. I guess Firebase uses Winston for logging somewhere and the change in node meant it didn't work properly? Sorry I haven't worked with node.js much myself.

But I think this issue can be closed.
